local-server/local-server-services/src/main/java/com/esri/samples/localserver/LocalServerServicesController.java

```diff
@@ -0,0 +1,287 @@
+        int servicesStarted = services.size();
+        for (LocalService service : services) {
+          if (service.getStatus() == LocalServerStatus.STOPPED) {
+            servicesStarted--;
+          }
+        }
+        if (servicesStarted == 0) {
+          System.out.println("Service Stopped");
+          stopServer();
+        }
+      }
+    });
+    localService.startAsync();
+  }
+
+  /**
+   * Gets the Local Service that is selected from the list of running services and stops it. 
+   * 
+   * @param event the event that fired this action
+   */
+  @FXML
+  protected void handleStopService(ActionEvent event) {
+
+    int serviceIndex = runningServices.getSelectionModel().getSelectedIndex();
+    services.get(serviceIndex).stopAsync();
+    runningServices.getItems().remove(serviceIndex);
+  }
+
+  /**
+   * Enables the URL and Stop Service button when a service is selected.
+   */
+  @FXML
+  protected void enableServiceButtons() {
+
+    int listIndex = runningServices.getSelectionModel().getSelectedIndex();
+    if (listIndex != -1) {
+      btnStopService.setDisable(false);
+      btnURL.setDisable(false);
+    } else {
+      btnStopService.setDisable(true);
+      btnURL.setDisable(true);
+    }
+  }
+
+  /**
+   * Opens a browser to the Local Service that is selected in the list of running services. 
+   * 
+   * @param event event that fired this action
+   */
+  @FXML
+  protected void handleURL(ActionEvent event) {
+    String url = runningServices.getSelectionModel().getSelectedItem().split(">")[1].trim();
+    hostServices.showDocument(url);
+  }
+
+  /**
+   * Stops any Local Services that are currently running and then stops the Local Server. 
+   */
+  void terminate() {
+    isServerStopping = true;
+    stopServicesAndServer();
+  }
+
+  /**
+   * Waits for all services that are running on the Local Server to completely stop and then stops the Local Server.
+   */
+  private void stopServicesAndServer() {
+
+    if (services != null && services.size() > 0) {
+      // stop any services that have been started
+      for (LocalService service : services) {
+        if (service.getStatus() == LocalServerStatus.STARTED) {
+
+          ListenableFuture<Void> stop = service.stopAsync();
+          // stop server once all services have stopped
+          //                    stop.addDoneListener(() -> {
+          //                      int servicesStarted = services.size();
+          //                      for (LocalService s : services) {
+          //                        if (s.getStatus() == LocalServerStatus.STOPPED) {
+          //                          servicesStarted--;
+          //                        }
+          //                      }
+          //                      if (servicesStarted == 0) {
+          //                        stopServer();
+          //                      }
+          //                    });

```

Comment:
@colinanderson 
My goal here is to stop any services that are running and then stop the local server if the stop local server button is hit or if the person exits out of the application.

I thought the commented out code would be the best way to achieve this, but it doesn't work.
If the person hits the stop local server button then the commented out code above works, services and server are stopped.
If person exits the application the services are stopped but the done listener is never fired to stop the server.

Seeing if you know a better way? Will comment above on a change that I made for now that has it working.